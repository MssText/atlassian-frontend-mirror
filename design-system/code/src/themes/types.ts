export type CodeTheme = {
  backgroundColor?: string | number;
  textColor?: string | number;
  substringColor?: string | number;
  keywordColor?: string | number;
  attributeColor?: string | number;
  selectorTagColor?: string | number;
  docTagColor?: string | number;
  nameColor?: string | number;
  builtInColor?: string | number;
  literalColor?: string | number;
  bulletColor?: string | number;
  codeColor?: string | number;
  additionColor?: string | number;
  regexpColor?: string | number;
  symbolColor?: string | number;
  variableColor?: string | number;
  templateVariableColor?: string | number;
  linkColor?: string | number;
  selectorAttributeColor?: string | number;
  selectorPseudoColor?: string | number;
  typeColor?: string | number;
  stringColor?: string | number;
  selectorIdColor?: string | number;
  selectorClassColor?: string | number;
  quoteColor?: string | number;
  templateTagColor?: string | number;
  deletionColor?: string | number;
  titleColor?: string | number;
  sectionColor?: string | number;
  commentColor?: string | number;
  metaKeywordColor?: string | number;
  metaColor?: string | number;
  functionColor?: string | number;
  numberColor?: string | number;
};

export interface CodeBlockTheme extends CodeTheme {
  lineNumberColor?: string | number;
  lineNumberBgColor?: string | number;
}
