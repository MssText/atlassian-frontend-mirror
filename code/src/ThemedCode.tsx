import React, { FC } from 'react';
import { withTheme, ThemeProvider } from 'styled-components';

import Code, { CodeProps } from './components/Code';

const CodeWithTheme = withTheme(Code);
const emptyTheme = {};

const ThemedCode: FC<CodeProps> = props => (
  <ThemeProvider theme={emptyTheme}>
    <CodeWithTheme {...props} />
  </ThemeProvider>
);

export default ThemedCode;
