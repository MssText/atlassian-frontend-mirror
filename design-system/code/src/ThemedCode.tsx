import React, { FC } from 'react';

import { ThemeProvider, withTheme } from 'styled-components';

import Code, { CodeProps } from './components/Code';

const CodeWithTheme = withTheme(Code);
const emptyTheme = {};

const ThemedCode: FC<CodeProps> = props => (
  <ThemeProvider theme={emptyTheme}>
    <CodeWithTheme {...props} />
  </ThemeProvider>
);

export default ThemedCode;
