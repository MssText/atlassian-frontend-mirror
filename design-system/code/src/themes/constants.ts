export const CODE_FONT_SIZE = 12;
export const CODE_LINE_HEIGHT = '20px';

// css vars used to decrease font family being repeatedly declared inline
export const VAR_CODE_FONT_FAMILY = '--code-font-family';
export const VAR_CODE_FONT_FAMILY_ITALIC = `${VAR_CODE_FONT_FAMILY}-italic`;
export const VAR_CODE_LINE_NUMBER_BG_COLOR = '--line-number-bg-color';
