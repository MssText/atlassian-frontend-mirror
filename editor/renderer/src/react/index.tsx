import React from 'react';
import { ComponentType } from 'react';
import { Fragment, Mark, MarkType, Node, Schema } from 'prosemirror-model';
import { Serializer } from '../';
import { RendererAppearance, StickyHeaderConfig } from '../ui/Renderer/types';
import { AnalyticsEventPayload } from '../analytics/events';

import {
  Doc,
  mergeTextNodes,
  isTextWrapper,
  TextWrapper,
  toReact,
} from './nodes';

import { toReact as markToReact } from './marks';
import {
  ProviderFactory,
  getMarksByOrder,
  isSameMark,
  EventHandlers,
  ExtensionHandlers,
  calcTableColumnWidths,
} from '@atlaskit/editor-common';
import { getText } from '../utils';

export interface RendererContext {
  objectAri?: string;
  containerAri?: string;
  adDoc?: any;
  schema?: Schema;
}

export interface ReactSerializerInit {
  providers?: ProviderFactory;
  eventHandlers?: EventHandlers;
  extensionHandlers?: ExtensionHandlers;
  portal?: HTMLElement;
  objectContext?: RendererContext;
  appearance?: RendererAppearance;
  disableHeadingIDs?: boolean;
  disableActions?: boolean;
  allowDynamicTextSizing?: boolean;
  allowHeadingAnchorLinks?: boolean;
  allowColumnSorting?: boolean;
  fireAnalyticsEvent?: (event: AnalyticsEventPayload) => void;
  shouldOpenMediaViewer?: boolean;
  allowAltTextOnImages?: boolean;
  stickyHeaders?: StickyHeaderConfig;
  allowMediaLinking?: boolean;
}

interface ParentInfo {
  parentIsIncompleteTask: boolean;
  path: Array<Node>;
}

interface FragmentChildContext {
  parentInfo?: ParentInfo;
  index: number;
}

type MarkWithContent = Partial<Mark<any>> & {
  content: Array<MarkWithContent | Node<any>>;
};

function mergeMarks(marksAndNodes: Array<MarkWithContent | Node>) {
  return marksAndNodes.reduce((acc, markOrNode) => {
    const prev = (acc.length && acc[acc.length - 1]) || null;

    if (
      markOrNode.type instanceof MarkType &&
      prev &&
      prev.type instanceof MarkType &&
      Array.isArray(prev.content) &&
      isSameMark(prev as Mark, markOrNode as Mark)
    ) {
      prev.content = mergeMarks(
        prev.content.concat((markOrNode as MarkWithContent).content),
      );
    } else {
      acc.push(markOrNode);
    }

    return acc;
  }, [] as Array<MarkWithContent | Node>);
}

export default class ReactSerializer implements Serializer<JSX.Element> {
  private providers?: ProviderFactory;
  private eventHandlers?: EventHandlers;
  private extensionHandlers?: ExtensionHandlers;
  private portal?: HTMLElement;
  private rendererContext?: RendererContext;
  private appearance?: RendererAppearance;
  private disableHeadingIDs?: boolean;
  private disableActions?: boolean;
  private headingIds: string[] = [];
  private allowDynamicTextSizing?: boolean;
  private allowHeadingAnchorLinks?: boolean;
  private allowColumnSorting?: boolean;
  private fireAnalyticsEvent?: (event: AnalyticsEventPayload) => void;
  private shouldOpenMediaViewer?: boolean;
  private allowAltTextOnImages?: boolean;
  private stickyHeaders?: StickyHeaderConfig;
  private allowMediaLinking?: boolean;

  constructor(init: ReactSerializerInit) {
    this.providers = init.providers;
    this.eventHandlers = init.eventHandlers;
    this.extensionHandlers = init.extensionHandlers;
    this.portal = init.portal;
    this.rendererContext = init.objectContext;
    this.appearance = init.appearance;
    this.disableHeadingIDs = init.disableHeadingIDs;
    this.disableActions = init.disableActions;
    this.allowDynamicTextSizing = init.allowDynamicTextSizing;
    this.allowHeadingAnchorLinks = init.allowHeadingAnchorLinks;
    this.allowColumnSorting = init.allowColumnSorting;
    this.fireAnalyticsEvent = init.fireAnalyticsEvent;
    this.shouldOpenMediaViewer = init.shouldOpenMediaViewer;
    this.allowAltTextOnImages = init.allowAltTextOnImages;
    this.stickyHeaders = init.stickyHeaders;
    this.allowMediaLinking = init.allowMediaLinking;
  }

  private resetState() {
    this.headingIds = [];
  }

  private getNodeProps(node: Node, parentInfo?: ParentInfo) {
    const path = parentInfo ? parentInfo.path : undefined;

    switch (node.type.name) {
      case 'date':
        return this.getDateProps(node, parentInfo);
      case 'hardBreak':
        return this.getHardBreakProps(node, path);
      case 'heading':
        return this.getHeadingProps(node, path);
      case 'media':
        return this.getMediaProps(node);
      case 'mediaSingle':
        return this.getMediaSingleProps(node, path);
      case 'table':
        return this.getTableProps(node);
      case 'tableHeader':
      case 'tableRow':
        return this.getTableChildrenProps(node);
      case 'taskItem':
        return this.getTaskItemProps(node);
      default:
        return this.getProps(node);
    }
  }

  serializeFragment(
    fragment: Fragment,
    props: any = {},
    target: any = Doc,
    key: string = 'root-0',
    parentInfo?: ParentInfo,
  ): JSX.Element | null {
    // This makes sure that we reset internal state on re-render.
    if (key === 'root-0') {
      this.resetState();
    }

    return this.renderNode(
      target,
      props,
      key,
      ReactSerializer.getChildNodes(fragment).map((node, index) => {
        if (isTextWrapper(node)) {
          return this.serializeTextWrapper(node.content);
        }
        return this.serializeFragmentChild(node, { index, parentInfo });
      }),
    );
  }

  private serializeFragmentChild = (
    node: Node,
    { index, parentInfo }: FragmentChildContext,
  ) => {
    const currentPath = (parentInfo && parentInfo.path) || [];

    const parentIsIncompleteTask =
      node.type.name === 'taskItem' && node.attrs.state !== 'DONE';

    const serializedContent = this.serializeFragment(
      node.content,
      this.getNodeProps(node, parentInfo),
      toReact(node),
      `${node.type.name}-${index}`,
      {
        parentIsIncompleteTask,
        path: [...currentPath, node],
      },
    );

    const marks = node.marks ? [...node.marks] : [];
    const isMediaSingle = node.type.name === 'mediaSingle';

    const getMarkProps = isMediaSingle
      ? this.getMediaMarkProps
      : this.getMarkProps;

    const shouldSkipMark = (mark: Mark): boolean =>
      this.allowMediaLinking !== true &&
      isMediaSingle &&
      mark.type.name === 'link';

    return marks.reverse().reduce((content, mark) => {
      if (shouldSkipMark(mark)) {
        return content;
      }

      return this.renderMark(
        markToReact(mark),
        getMarkProps(mark),
        `${mark.type.name}-${index}`,
        content,
      );
    }, serializedContent);
  };

  private getMediaMarkProps = (mark: Mark) =>
    mark.type.name === 'link'
      ? {
          ...this.getMarkProps(mark),
          isMediaLink: true,
        }
      : this.getMarkProps(mark);

  private serializeTextWrapper(content: Node[]) {
    return ReactSerializer.buildMarkStructure(content).map((mark, index) =>
      this.serializeMark(mark, index),
    );
  }

  private serializeMark(mark: Mark, index: number = 0) {
    if (mark.type.name === 'text') {
      return (mark as any).text;
    }

    const content = (
      (mark as any).content || []
    ).map((child: Mark, index: number) => this.serializeMark(child, index));
    return this.renderMark(
      markToReact(mark),
      this.getMarkProps(mark),
      `${mark.type.name}-${index}`,
      content,
    );
  }

  private renderNode(
    NodeComponent: ComponentType<any>,
    props: any,
    key: string,
    content: string | JSX.Element | any[] | null | undefined,
  ): JSX.Element {
    return (
      <NodeComponent key={key} {...props}>
        {content}
      </NodeComponent>
    );
  }

  private renderMark(
    MarkComponent: ComponentType<any>,
    props: any,
    key: string,
    content: any,
  ) {
    return (
      <MarkComponent key={key} {...props}>
        {content}
      </MarkComponent>
    );
  }

  private getTableChildrenProps(node: Node) {
    return {
      ...this.getProps(node),
      allowColumnSorting: this.allowColumnSorting,
    };
  }

  private getTableProps(node: Node) {
    return {
      ...this.getProps(node),
      allowColumnSorting: this.allowColumnSorting,
      columnWidths: calcTableColumnWidths(node),
      tableNode: node,
      stickyHeaders: this.stickyHeaders,
    };
  }

  private getDateProps(
    node: Node,
    parentInfo: { parentIsIncompleteTask: boolean } | undefined,
  ) {
    return {
      timestamp: node.attrs && node.attrs.timestamp,
      parentIsIncompleteTask: parentInfo && parentInfo.parentIsIncompleteTask,
    };
  }

  private getMediaSingleProps(node: Node, path: Array<Node> = []) {
    const {
      nodes: { expand, nestedExpand, layoutColumn },
      marks: { link },
    } = node.type.schema;
    const blockNodeNames = [expand, nestedExpand, layoutColumn]
      .filter(node => Boolean(node)) // Check if the node exist first
      .map(node => node.name);
    const isInsideOfBlockNode =
      path &&
      path.some(n => n.type && blockNodeNames.indexOf(n.type.name) > -1);

    const isLinkMark = (mark: Mark) => mark.type === link;

    return {
      ...this.getProps(node),
      marks: node.marks.filter(
        m => !isLinkMark(m) || this.allowMediaLinking === true,
      ),
      isLinkMark,
      isInsideOfBlockNode,
    };
  }

  private getMediaProps(node: Node) {
    return {
      ...this.getProps(node),
      shouldOpenMediaViewer: this.shouldOpenMediaViewer,
      allowAltTextOnImages: this.allowAltTextOnImages,
    };
  }

  private getTaskItemProps(node: Node) {
    return {
      ...this.getProps(node),
      disabled: this.disableActions,
    };
  }

  private getHardBreakProps(node: Node, path: Array<Node> = []) {
    let forceNewLine = false;

    const parentNode: Node | null =
      path.length > 0 ? path[path.length - 1] : null;

    if (parentNode && parentNode.lastChild === node) {
      forceNewLine = true;
    }

    return {
      ...this.getProps(node),
      forceNewLine,
    };
  }

  private getProps(node: Node) {
    return {
      text: node.text,
      providers: this.providers,
      eventHandlers: this.eventHandlers,
      extensionHandlers: this.extensionHandlers,
      portal: this.portal,
      rendererContext: this.rendererContext,
      serializer: this,
      content: node.content ? node.content.toJSON() : undefined,
      allowDynamicTextSizing: this.allowDynamicTextSizing,
      allowHeadingAnchorLinks: this.allowHeadingAnchorLinks,
      rendererAppearance: this.appearance,
      fireAnalyticsEvent: this.fireAnalyticsEvent,
      nodeType: node.type.name,
      marks: node.marks,
      ...node.attrs,
    };
  }

  private headingAnchorSupported(path: Array<Node> = []): boolean {
    return (
      path.length === 0 || path[path.length - 1].type.name === 'layoutColumn'
    );
  }

  private getHeadingProps(node: Node, path: Array<Node> = []) {
    return {
      ...node.attrs,
      content: node.content ? node.content.toJSON() : undefined,
      headingId: this.getHeadingId(node),
      showAnchorLink:
        this.allowHeadingAnchorLinks &&
        !this.disableHeadingIDs &&
        this.headingAnchorSupported(path),
    };
  }

  // The return value of this function is NOT url encoded,
  // In HTML5 standard, id can contain any characters, encoding is no necessary.
  // Plus we trying to avoid double encoding, therefore we leave the value as is.
  // Remember to use encodeURIComponent when generating url from the id value.
  private getHeadingId(node: Node) {
    if (this.disableHeadingIDs || !node.content.size) {
      return;
    }

    // We are not use node.textContent here, because we would like to handle cases where
    // headings only contain inline blocks like emoji, status and date.
    const nodeContent = (node as any).content
      .toJSON()
      .reduce((acc: string, node: any) => acc.concat(getText(node) || ''), '')
      .trim()
      .replace(/\s/g, '-');

    if (!nodeContent) {
      return;
    }

    return this.getUniqueHeadingId(nodeContent);
  }

  private getUniqueHeadingId(baseId: string, counter = 0): string {
    if (counter === 0 && this.headingIds.indexOf(baseId) === -1) {
      this.headingIds.push(baseId);
      return baseId;
    } else if (counter !== 0) {
      const headingId = `${baseId}.${counter}`;
      if (this.headingIds.indexOf(headingId) === -1) {
        this.headingIds.push(headingId);
        return headingId;
      }
    }

    return this.getUniqueHeadingId(baseId, ++counter);
  }

  private getMarkProps = (mark: Mark): any => {
    const { key, ...otherAttrs } = mark.attrs;
    return {
      eventHandlers: this.eventHandlers,
      fireAnalyticsEvent: this.fireAnalyticsEvent,
      markKey: key,
      ...otherAttrs,
    };
  };

  static getChildNodes(fragment: Fragment): (Node | TextWrapper)[] {
    const children: Node[] = [];
    fragment.forEach(node => {
      children.push(node);
    });
    return mergeTextNodes(children) as Node[];
  }

  static getMarks(node: Node): Mark[] {
    if (!node.marks || node.marks.length === 0) {
      return [];
    }

    return getMarksByOrder(node.marks);
  }

  static buildMarkStructure(content: Node[]) {
    return mergeMarks(
      content.map(node => {
        const nodeMarks = this.getMarks(node);
        if (nodeMarks.length === 0) {
          return node;
        }

        return nodeMarks.reverse().reduce((acc, mark) => {
          const { eq } = mark;

          return {
            ...mark,
            eq,
            content: [acc],
          };
        }, node as any);
      }),
    ) as Mark[];
  }

  // TODO: ED-9004 Remove unused ReactSerializer.fromSchema in renderer
  // https://sourcegraph-frontend.internal.shared-prod.us-west-2.kitt-inf.net/search?q=ReactSerializer.fromSchema&patternType=literal
  static fromSchema(_: unknown, init: ReactSerializerInit) {
    if (process.env.NODE_ENV !== 'production') {
      // eslint-disable-next-line no-console
      console.warn(
        'ReactSerializer.fromSchema is deprecated. Please use the constructor instead via new ReactSerializer()',
      );
    }
    return new ReactSerializer(init);
  }
}
