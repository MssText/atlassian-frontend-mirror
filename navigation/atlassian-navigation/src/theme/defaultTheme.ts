import {
  B200,
  B400,
  B50,
  N0,
  N20,
  N200,
  N40,
  N600,
  N700,
} from '@atlaskit/theme/colors';
import { token } from '@atlaskit/tokens';

import { hexToRGBA } from './themeHelpers';
import { Mode } from './types';

const defaultTheme: { mode: Mode } = {
  mode: {
    create: {
      active: {
        color: token('color.text.onBold', N0),
        backgroundColor: token(
          'color.background.boldBrand.pressed',
          hexToRGBA(B400, 0.8),
        ),
        boxShadow: '',
      },
      default: {
        color: token('color.text.onBold', N0),
        backgroundColor: token('color.background.boldBrand.resting', B400),
        boxShadow: '',
      },
      focus: {
        color: token('color.text.onBold', N0),
        backgroundColor: token('color.background.boldBrand.resting', B400),
        boxShadow: `0 0 0 2px ${token(
          'color.border.focus',
          'rgb(128,169,230)',
        )}`,
      },
      hover: {
        color: token('color.text.onBold', N0),
        backgroundColor: token(
          'color.background.boldBrand.hover',
          hexToRGBA(B400, 0.9),
        ),
        boxShadow: '',
      },
      selected: { color: '', backgroundColor: '', boxShadow: '' },
    },
    iconButton: {
      active: {
        color: token('color.text.mediumEmphasis', B400),
        backgroundColor: token(
          'color.background.subtleNeutral.pressed',
          hexToRGBA(B50, 0.6),
        ),
        boxShadow: '',
      },
      default: {
        color: token('color.text.mediumEmphasis', N600),
        backgroundColor: 'transparent',
        boxShadow: '',
      },
      focus: {
        color: token('color.text.mediumEmphasis', N600),
        backgroundColor: token(
          'color.background.subtleBrand.hover',
          hexToRGBA(B50, 0.5),
        ),
        boxShadow: `0 0 0 2px ${token('color.border.focus', B200)}`,
      },
      hover: {
        color: token('color.text.mediumEmphasis', B400),
        backgroundColor: token(
          'color.background.subtleNeutral.hover',
          hexToRGBA(B50, 0.9),
        ),
        boxShadow: '',
      },
      selected: {
        color: token('color.text.selected', ''),
        backgroundColor: token('color.background.selected.resting', ''),
        boxShadow: '',
      },
    },
    navigation: {
      backgroundColor: token('color.background.default', N0),
      color: token('color.text.lowEmphasis', N200),
    },
    productHome: {
      backgroundColor: token('color.text.brand', B400),
      color: token('color.text.highEmphasis', N700),
      borderRight: `1px solid ${token(
        'color.border.neutral',
        hexToRGBA(N200, 0.3),
      )}`,
      // TODO: (DSP-1256) These colors should be moved into the Logo package
      // eslint-disable-next-line @atlaskit/design-system/ensure-design-token-usage
      iconGradientStart: B400,
      // eslint-disable-next-line @atlaskit/design-system/ensure-design-token-usage
      iconGradientStop: B200,
      // eslint-disable-next-line @atlaskit/design-system/ensure-design-token-usage
      iconColor: B200,
    },
    primaryButton: {
      active: {
        color: token('color.text.mediumEmphasis', B400),
        backgroundColor: token(
          'color.background.subtleNeutral.pressed',
          hexToRGBA(B50, 0.7),
        ),
        boxShadow: '',
      },
      default: {
        color: token('color.text.mediumEmphasis', N600),
        backgroundColor: 'transparent',
        boxShadow: '',
      },
      focus: {
        color: token('color.text.mediumEmphasis', N600),
        backgroundColor: '',
        boxShadow: `0 0 0 2px ${token('color.border.focus', B200)}`,
      },
      hover: {
        color: token('color.text.mediumEmphasis', B400),
        backgroundColor: token(
          'color.background.subtleNeutral.hover',
          hexToRGBA(B50, 0.9),
        ),
        boxShadow: '',
      },
      selected: {
        color: token('color.text.mediumEmphasis', B400),
        backgroundColor: token('color.background.selected.resting', ''),
        boxShadow: '',
        borderColor: token('color.iconBorder.brand', B400),
        bordorBottom: `4px solid ${token('color.iconBorder.brand', B400)}`,
      },
    },
    search: {
      default: {
        backgroundColor: token('color.background.default', N0),
        color: token('color.text.lowEmphasis', N200),
        borderColor: token('color.border.neutral', N40),
      },
      focus: {
        borderColor: token('color.border.focus', B200),
      },
      hover: {
        color: token('color.text.brand', B400),
        backgroundColor: token('color.background.default', hexToRGBA(B50, 0.9)),
      },
    },
    skeleton: {
      backgroundColor: token('color.background.subtleNeutral.resting', N20),
      opacity: 1,
    },
  },
};

export const DEFAULT_THEME_NAME = 'atlassian';
export default defaultTheme;
