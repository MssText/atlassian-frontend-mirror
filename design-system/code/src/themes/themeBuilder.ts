import {
  borderRadius,
  codeFontFamily,
  fontSize,
  gridSize,
} from '@atlaskit/theme/constants';
import { ThemeProps } from '@atlaskit/theme/types';

import { defaultColors } from './defaultTheme';
import { CodeBlockTheme, CodeTheme } from './types';

const codeContainerStyle = {
  fontFamily: codeFontFamily,
  fontSize: '12px',
  lineHeight: 20 / 12,
  padding: gridSize(),
};

const lineNumberContainerStyle = (theme: CodeBlockTheme) => ({
  fontSize: `${fontSize()}px`,
  lineHeight: 20 / 14,
  color: theme.lineNumberColor,
  backgroundColor: theme.lineNumberBgColor,
  flexShrink: 0,
  padding: gridSize(),
  textAlign: 'right',
  userSelect: 'none',
});

const sharedCodeStyle = (theme: CodeTheme) => ({
  key: {
    color: theme.keywordColor,
    fontWeight: 'bolder',
  },
  keyword: {
    color: theme.keywordColor,
    fontWeight: 'bolder',
  },
  'attr-name': {
    color: theme.attributeColor,
  },
  selector: {
    color: theme.selectorTagColor,
  },
  comment: {
    color: theme.commentColor,
    fontFamily: `SFMono-MediumItalic, ${codeFontFamily()}`,
    fontStyle: 'italic',
  },
  'block-comment': {
    color: theme.commentColor,
    fontFamily: `SFMono-MediumItalic, ${codeFontFamily()}`,
    fontStyle: 'italic',
  },
  'function-name': {
    color: theme.sectionColor,
  },
  'class-name': {
    color: theme.sectionColor,
  },
  doctype: {
    color: theme.docTagColor,
  },
  substr: {
    color: theme.substringColor,
  },
  namespace: {
    color: theme.nameColor,
  },
  builtin: {
    color: theme.builtInColor,
  },
  entity: {
    color: theme.literalColor,
  },
  bullet: {
    color: theme.bulletColor,
  },
  code: {
    color: theme.codeColor,
  },
  addition: {
    color: theme.additionColor,
  },
  regex: {
    color: theme.regexpColor,
  },
  symbol: {
    color: theme.symbolColor,
  },
  variable: {
    color: theme.variableColor,
  },
  url: {
    color: theme.linkColor,
  },
  'selector-attr': {
    color: theme.selectorAttributeColor,
  },
  'selector-pseudo': {
    color: theme.selectorPseudoColor,
  },
  type: {
    color: theme.typeColor,
  },
  string: {
    color: theme.stringColor,
  },
  quote: {
    color: theme.quoteColor,
  },
  tag: {
    color: theme.templateTagColor,
  },
  deletion: {
    color: theme.deletionColor,
  },
  title: {
    color: theme.titleColor,
  },
  section: {
    color: theme.sectionColor,
  },
  'meta-keyword': {
    color: theme.metaKeywordColor,
  },
  meta: {
    color: theme.metaColor,
  },
  italic: {
    fontStyle: 'italic',
  },
  bold: {
    fontWeight: 'bolder',
  },
  function: {
    color: theme.functionColor,
  },
  number: {
    color: theme.numberColor,
  },
});

const codeStyle = (theme: CodeTheme) => ({
  fontFamily: codeFontFamily,
  fontSize: '12px',
  background: theme.backgroundColor,
  color: theme.textColor,
  borderRadius: borderRadius(),
  display: 'flex',
  lineHeight: 20 / 12,
  overflowX: 'auto',
  whiteSpace: 'pre',
});

const codeBlockStyle = (theme: CodeBlockTheme) => ({
  'pre[class*="language-"]': codeStyle(theme),
  ...sharedCodeStyle(theme),
});

const inlineCodeStyle = (theme: CodeTheme) => ({
  'pre[class*="language-"]': {
    ...codeStyle(theme),
    padding: '2px 4px',
    display: 'inline',
    whiteSpace: 'pre-wrap',
  },
  ...sharedCodeStyle(theme),
});

export function applyTheme(
  theme: ThemeProps | CodeTheme | CodeBlockTheme = {},
) {
  const newTheme = { ...defaultColors(theme), ...theme };
  return {
    lineNumberContainerStyle: lineNumberContainerStyle(newTheme),
    codeBlockStyle: codeBlockStyle(newTheme),
    inlineCodeStyle: inlineCodeStyle(newTheme),
    codeContainerStyle,
  };
}
