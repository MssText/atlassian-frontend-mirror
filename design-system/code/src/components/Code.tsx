import React, { PureComponent, ReactNode } from 'react';

import { PrismAsyncLight as SyntaxHighlighter } from 'react-syntax-highlighter';

import { ThemeProps } from '@atlaskit/theme/types';

import { normalizeLanguage, SupportedLanguages } from '../supportedLanguages';
import { applyTheme } from '../themes/themeBuilder';
import { CodeTheme } from '../themes/types';

export interface CodeProps {
  /** The style object to apply to code */
  codeStyle?: {};
  /** The props that will be spread into the `<code>` tag. Useful for styling/assigning classNames */
  codeTagProps?: {};
  /** The language in which the code is written */
  language: SupportedLanguages | string;
  /** The element or custom react component to use in place of the default span tag */
  preTag?: ReactNode | string;
  /** The code to be formatted */
  text: string;
  /** A custom theme to be applied, implements the CodeTheme interface */
  theme?: CodeTheme | ThemeProps;
}

export default class Code extends PureComponent<CodeProps, {}> {
  static defaultProps = {
    theme: {},
    codeTagProps: {},
    preTag: 'span',
  };

  render() {
    const { inlineCodeStyle } = applyTheme(this.props.theme);
    const language = normalizeLanguage(this.props.language);

    const props = {
      language,
      PreTag: this.props.preTag,
      style: this.props.codeStyle || inlineCodeStyle,
      codeTagProps: this.props.codeTagProps,
    };

    return <SyntaxHighlighter {...props}>{this.props.text}</SyntaxHighlighter>;
  }
}
