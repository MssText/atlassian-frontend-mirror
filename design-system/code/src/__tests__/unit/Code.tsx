import React from 'react';

import { cleanup, render } from '@testing-library/react';

import Code from '../../components/Code';
import ThemedCode from '../../ThemedCode';
import { defaultColors } from '../../themes/defaultTheme';

const jsCode = `const map = new Map({ key: 'value' })`;
const jsTestId = 'jstestid';

const javaCode = `public class HelloWorld
{
	public static void main(String[] args) {
		System.out.println("Hello World!");
	}
}`;

const pyCode = `def transform_data(data_frame, rolling_value):
rolling_df = pd.DataFrame(data_frame)
return rolling_df.rolling(rolling_value, min_periods=1, center=True).mean()`;

describe('Code', () => {
  const codeJavascript = (
    <Code text={jsCode} testId={jsTestId} language="javascript" />
  );
  const codeLanguageNotSupported = (
    <Code text={jsCode} testId="not-supported" language="dde" />
  );
  afterEach(cleanup);
  test('should render with language javascript', () => {
    const { getByTestId } = render(codeJavascript);
    expect(getByTestId(jsTestId)).toBeDefined();
    expect(getByTestId(jsTestId).getAttribute('data-code-lang')).toEqual(
      'javascript',
    );
  });
  test('should render with language that is not supported', () => {
    expect(
      render(codeLanguageNotSupported).getByTestId('not-supported'),
    ).toBeDefined();
  });

  describe('theming', () => {
    it('should apply theme bg color (dark)', () => {
      const { getByTestId } = render(
        <ThemedCode
          text={javaCode}
          language="java"
          testId="theme"
          theme={{ mode: 'dark' }}
        />,
      );
      expect(getByTestId('theme')).toHaveStyle(
        `background-color:
        ${defaultColors({ mode: 'dark' }).backgroundColor};`,
      );
    });

    it('should apply theme bg color', () => {
      const { getByTestId } = render(
        <ThemedCode text={javaCode} language="java" testId="theme" />,
      );
      expect(getByTestId('theme')).toHaveStyle(
        `background-color: ${
          defaultColors({ mode: 'light' }).backgroundColor
        };`,
      );
    });
  });

  test('should render a div instead of a span', () => {
    const { getByTestId } = render(
      <ThemedCode
        preTag="div"
        testId="python"
        text={pyCode}
        language="python"
      />,
    );

    expect(getByTestId('python').nodeName).toEqual('DIV');
  });
  test('should style code container with a red color', () => {
    const { container } = render(
      <ThemedCode
        preTag="div"
        codeTagProps={{ style: { color: 'red' } }}
        text={pyCode}
        language="python"
      />,
    );
    expect(container.querySelector('code')).toHaveStyle('color: red;');
  });
  test('should style numbers', () => {
    const { getByText } = render(
      <ThemedCode
        testId="test"
        codeStyle={{ number: { color: 'red' } }}
        text="x = 10 + 12"
        language="python"
      />,
    );
    expect(getByText('10')).toHaveStyle('number: red;');
    expect(getByText('12')).toHaveStyle('number: red;');
  });
});
