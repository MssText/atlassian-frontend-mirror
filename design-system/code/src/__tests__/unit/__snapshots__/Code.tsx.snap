// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Code should apply style to code 1`] = `
<ThemeProvider
  theme={Object {}}
>
  <WithTheme(Code)
    codeStyle={
      Object {
        "style": Object {
          "color": "blue",
        },
      }
    }
    language="python"
    text="def transform_data(data_frame, rolling_value):
rolling_df = pd.DataFrame(data_frame)
return rolling_df.rolling(rolling_value, min_periods=1, center=True).mean()"
  />
</ThemeProvider>
`;

exports[`Code should render a div with a red color 1`] = `
<ThemeProvider
  theme={Object {}}
>
  <WithTheme(Code)
    codeTagProps={
      Object {
        "style": Object {
          "color": "red",
        },
      }
    }
    language="python"
    preTag="div"
    text="def transform_data(data_frame, rolling_value):
rolling_df = pd.DataFrame(data_frame)
return rolling_df.rolling(rolling_value, min_periods=1, center=True).mean()"
  />
</ThemeProvider>
`;
